specreduce/wavelength_calibration.py

```diff
@@ -0,0 +1,303 @@
+
+        if fitter is None:
+            if self.model.linear:
+                self.fitter = LinearLSQFitter(calc_uncertainties=True)
+            else:
+                self.fitter = LMLSQFitter(calc_uncertainties=True)
+        else:
+            self.fitter = fitter
+
+        if self.default_refinement_method in ("gaussian", "min", "max"):
+            if 'range' not in self.default_refinement_kwargs:
+                # We may want to adjust this default based on real calibration spectra
+                self.default_refinement_method['range'] = 10
+
+        self._lines = []
+        if isinstance(lines, str):
+            if lines in self.available_line_lists:
+                raise ValueError(f"Line list '{lines}' is not an available line list.")
+        else:
+            for line in lines:
+                if isinstance(line, CalibrationLine):
+                    self._lines.append(line)
+                else:
+                    self._lines.append(CalibrationLine(self.input_spectrum, line[0], line[1],
+                                       self.default_refinement_method,
+                                       self.default_refinement_kwargs))
+
+    def _clear_cache(self, *attrs):
+        """
+        provide convenience function to clearing the cache for cached_properties
+        """
+        if not len(attrs):
+            attrs = self._cached_properties
+        for attr in attrs:
+            if attr in self.__dict__:
+                del self.__dict__[attr]
+
+    @property
+    def input_spectrum(self):
+        return self._input_spectrum
+
+    @input_spectrum.setter
+    def input_spectrum(self, new_spectrum):
+        # We want to clear the refined locations if a new calibration spectrum is provided
+        self._clear_cache()
+        for line in self.lines:
+            line.input_spectrum = new_spectrum
+        self._input_spectrum = new_spectrum
+
+    @property
+    def lines(self):
+        return self._lines
+
+    @lines.setter
+    def lines(self, new_lines):
+        self._clear_cache()
+        self._lines = new_lines
+
+    @property
+    def model(self):
+        return self._model
+
+    @model.setter
+    def model(self, new_model):
+        self._clear_cache()
+        self._model = new_model
+
+    @property
+    def refined_lines(self):
+        return [line.with_refined_pixel for line in self.lines]
+
+    @property
+    def refined_pixels(self):
+        # useful for plotting over spectrum to change input lines/refinement options
+        return [(line.wavelength, line.refined_pixel) for line in self.lines]
+
+    @cached_property
+    def wcs(self):
+        # computes and returns WCS after fitting self.model to self.refined_pixels
+        x = [line[1] for line in self.refined_pixels] * u.pix
+        y = [line[0].value for line in self.refined_pixels] * self.spectral_unit
+
+        # Fit the model
+        self._model = self.fitter(self._model, x, y)
```

Comment:
does this prevent successive fitting calls?

Reply:
I don't think so...I guess successive calls would be starting from the previous fitted parameter values, but I don't think it prevents fitting. Do you think having a separate `_fitted_model` parameter would be better?

Reply:
I'm not sure 🤷, I think the way you have it - assuming the output can be passed as the input - is both quite handy but also maybe unexpected behavior unless we document carefully.